### PR TITLE
castxml: unbreak binary by adding llvm to buildInputs

### LINKED ISSUE #318055: `castxml` binary is broken as it does not correctly link libLLVM
### Describe the bug

Sometime between 23.11 and now, the `castxml` binary broke due to a linking issue.

### Steps To Reproduce
```
$ nix run nixpkgs/release-23.11#castxml  # succeeds
$ nix run nixpkgs#castxml
/nix/store/ccz3hkgbf2fi5rylc2xw6iw1ks83lfqf-castxml-0.6.6/bin/castxml: error while loading shared libraries: libLLVM-17.so: cannot open shared object file: No such file or directory
$ nix shell nixpkgs#castxml --command which castxml | xargs ldd
        linux-vdso.so.1 (0x00007ffea9d1e000)
        libclang-cpp.so.17 => /nix/store/wxxgsgjxbnkkyczgf8lkbfrsqiywm8bi-clang-17.0.6-lib/lib/libclang-cpp.so.17 (0x00007f0175a00000)
        libLLVM-17.so => not found
        libstdc++.so.6 => /nix/store/xvzz97yk73hw03v5dhhz3j47ggwf1yq1-gcc-13.2.0-lib/lib/libstdc++.so.6 (0x00007f0175600000)
        libm.so.6 => /nix/store/k7zgvzp2r31zkg9xqgjim7mbknryv6bs-glibc-2.39-52/lib/libm.so.6 (0x00007f01798aa000)
        libgcc_s.so.1 => /nix/store/xvzz97yk73hw03v5dhhz3j47ggwf1yq1-gcc-13.2.0-lib/lib/libgcc_s.so.1 (0x00007f0179885000)
        libc.so.6 => /nix/store/k7zgvzp2r31zkg9xqgjim7mbknryv6bs-glibc-2.39-52/lib/libc.so.6 (0x00007f0175413000)
        libLLVM-17.so => /nix/store/ax9wbqknwawnl21kvkp8l6jn2r6q0l90-llvm-17.0.6-lib/lib/libLLVM-17.so (0x00007f016ca00000)
        /nix/store/k7zgvzp2r31zkg9xqgjim7mbknryv6bs-glibc-2.39-52/lib/ld-linux-x86-64.so.2 => /nix/store/k7zgvzp2r31zkg9xqgjim7mbknryv6bs-glibc-2.39-52/lib64/ld-linux-x86-64.so.2 (0x00007f017998f000)
        libffi.so.8 => /nix/store/kbg5m7fyi1w23fyfmxjhhzcbd577rpg0-libffi-3.4.6/lib/libffi.so.8 (0x00007f0179872000)
        librt.so.1 => /nix/store/k7zgvzp2r31zkg9xqgjim7mbknryv6bs-glibc-2.39-52/lib/librt.so.1 (0x00007f017986b000)
        libdl.so.2 => /nix/store/k7zgvzp2r31zkg9xqgjim7mbknryv6bs-glibc-2.39-52/lib/libdl.so.2 (0x00007f0179866000)
        libz.so.1 => /nix/store/lv6nackqis28gg7l2ic43f6nk52hb39g-zlib-1.3.1/lib/libz.so.1 (0x00007f0179848000)
        libncursesw.so.6 => /nix/store/j4m3lwhh4x0bzilvi77d512ahry775nq-ncurses-6.4/lib/libncursesw.so.6 (0x00007f01797d4000)
        libxml2.so.2 => /nix/store/kf7fzax0av9him9m4zcmbqmihh5108wq-libxml2-2.12.7/lib/libxml2.so.2 (0x00007f01758a0000)
```

### Notify maintainers

@AndersonTorres

### PR DESCRIPTION
**Pull request by bcdarwin**
Adds LLVM to `buildInputs`.

Also sets `CLANG_RESOURCE_DIR` correctly and enables tests.

Fixes #318055.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by paparodeo — APPROVED** [review]
LGTM

could add testers to the argument set and
```nix
  passthru.tests = testers.testVersion {
    package = finalAttrs.finalPackage;
  };
```
to prevent from happening again.

File: pkgs/by-name/ca/castxml/package.nix [hunk 1]
```diff
@@ -66,6 +66,10 @@ stdenv.mkDerivation (finalAttrs: {
     runHook postCheck
   '';
 
+  passthru.tests = testers.testVersion {
```
Comment (paparodeo) [inline pkgs/by-name/ca/castxml/package.nix:69]: need to add `testers` as an input at the top of the file.

Comment (bcdarwin) [reply pkgs/by-name/ca/castxml/package.nix:70]: Yep, added.

**Review by paparodeo — APPROVED** [review]
LGTM: `nix-build -A castxml.tests` succeeds on x64 linux/darwin

File: pkgs/by-name/ca/castxml/package.nix [hunk 2]
```diff
@@ -40,6 +40,7 @@ stdenv.mkDerivation (finalAttrs: {
     libffi
     libxml2
     zlib
+    llvm
```
Comment (AndersonTorres) [inline pkgs/by-name/ca/castxml/package.nix:43]:
```suggestion
    llvm
    zlib
```

Alphabetical order, since order is relevant in lists.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ca/castxml/package.nix (hunk 2 above)
Comment (bcdarwin) [reply pkgs/by-name/ca/castxml/package.nix]: Done.

**Review by AndersonTorres — APPROVED** [review]

**Comment by bcdarwin** [comment]
Added another commit to unbreak the test suite.

File: pkgs/by-name/ca/castxml/package.nix [hunk 3]
```diff
@@ -39,32 +39,28 @@ stdenv.mkDerivation (finalAttrs: {
...
     libxml2
+    llvm
     zlib
   ] ++ lib.optionals (!stdenv.isDarwin) [
     libclang
   ];
 
-  cmakeFlags = [
-    (lib.cmakeOptionType "path" "CLANG_RESOURCE_DIR" "${lib.getDev libclang}")
+  cmakeFlags = let clangVersionMajor = builtins.elemAt (lib.strings.splitString "." libclang.version) 0; in [
```
Comment (AndersonTorres) [inline pkgs/by-name/ca/castxml/package.nix:48]:
```suggestion
  cmakeFlags = let
    clangVersionMajor = builtins.elemAt (lib.strings.splitString "." libclang.version) 0; 
  in [
```

P.S.: isn't better to use `lib.major`?

https://ryantm.github.io/nixpkgs/functions/library/versions/#function-library-lib.versions.major

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ca/castxml/package.nix (hunk 3 above)
Comment (AndersonTorres) [reply pkgs/by-name/ca/castxml/package.nix:48]: P.P.S.: there is few to no need of using blocks of indirectio,  when the function is already on `lib` scope.

Comment (bcdarwin) [reply pkgs/by-name/ca/castxml/package.nix]: Done.

**Comment by bcdarwin** [comment]
Result of `nixpkgs-review pr 318140` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>castxml</li>
    <li>gnss-sdr</li>
    <li>gnuradio</li>
    <li>gnuradioPackages.osmosdr</li>
    <li>gnuradioPackages.osmosdr.dev</li>
    <li>ns-3</li>
    <li>python311Packages.pygccxml</li>
    <li>python311Packages.pygccxml.dist</li>
    <li>python312Packages.pygccxml</li>
    <li>python312Packages.pygccxml.dist</li>
  </ul>
</details>

**Review by leona-ya — APPROVED** [review]

**Pull request merged**